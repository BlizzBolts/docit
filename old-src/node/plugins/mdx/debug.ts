import type { Plugin } from "unified";
import type { ResolvedUserConfig } from "../../types.js";

export const debug = (config: ResolvedUserConfig): Plugin => {
  return (ast) => {};
};
