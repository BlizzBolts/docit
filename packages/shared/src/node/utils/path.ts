import path from "node:path";
import url from "node:url";
/**
 * isomorphic __dirname getter
 *
 * @param {import.meta.url} importMetaUrl
 * @returns
 */
export const getDirname = (importMetaUrl: string) => {
  return path.dirname(url.fileURLToPath(importMetaUrl));
};
