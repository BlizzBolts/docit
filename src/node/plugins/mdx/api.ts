import { visit } from "unist-util-visit";
import { Plugin } from "unified";
import { Link, Text } from "mdast";
import { ResolvedUserConfig } from "../../types.js";
import { parseMdxToTree } from "../../utils/index.js";
import { select } from "unist-util-select";
import resolve from "resolve";
const { sync } = resolve;

export const api = (config: ResolvedUserConfig): Plugin => {
  return (ast) => {
    visit(ast, "link", (link: Link, _, parent) => {
      const text = select("text", link) as Text;

      if (!link || !text) {
        return;
      }

      if (text.value.toLowerCase() === "props" && link.url) {
        const componentPath = sync(link.url, {
          basedir: config.docs,
        });
        const result = parseMdxToTree(`
<ApiTable get={() => import('${componentPath}?needParse')} path="${componentPath}"/>
       `);
        const index = parent.children.findIndex((o) => o === link);
        parent.children.splice(index, 1, result);
      }
    });
  };
};
