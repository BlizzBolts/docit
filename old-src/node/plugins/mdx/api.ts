import path from "path";
import resolve from "resolve";
import { select } from "unist-util-select";
import { visit } from "unist-util-visit";
import type { Plugin } from "unified";
import type { Link, Text } from "mdast";

import type { VFile } from "vfile";
import { parseMdxToTree } from "../../utils/index.js";
import type { ResolvedUserConfig } from "../../types";

const { sync } = resolve;

export const api = (config: ResolvedUserConfig): Plugin => {
  return (ast, file: VFile) => {
    visit(ast, "link", (link: Link, _, parent) => {
      const text = select("text", link) as Text;

      if (!link || !text) {
        return;
      }

      if (text.value.toLowerCase() === "props" && link.url) {
        const componentPath = sync(link.url, {
          basedir: path.dirname(file.path),
        });
        const result = parseMdxToTree(`
<ApiTable get={() => import('${componentPath}?needParse')} path="${componentPath}"/>
       `);
        const index = parent.children.findIndex((o) => o === link);
        parent.children.splice(index, 1, result);
      }
    });
  };
};
