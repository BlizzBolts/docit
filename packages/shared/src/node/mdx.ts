import remarkFrontmatter from "remark-frontmatter";
import remarkParse from "remark-parse";
import remarkStringify from "remark-stringify";
import { unified } from "unified";
import type { Node } from "unified/lib";
import { matter } from "vfile-matter";
import type { VFile } from "vfile";
import { glob } from "glob";
import fsx from "fs-extra";
import { markdownPathToRoutePath } from "../shared/utils/markdown";

const matterParser = () => {
  return (tree: Node, file: VFile) => {
    matter(file);
  };
};
export const getFrontMatter = async (mdxContent: string) => {
  const file = await unified()
    .use(remarkParse)
    .use(remarkStringify)
    .use(remarkFrontmatter, ["yaml"])
    .use(matterParser)
    .process(mdxContent);

  return file.data.matter;
};

export const scanMarkdowns = async (docRoot: string) => {
  const docPaths = await glob("./**/*.{md,mdx}", {
    cwd: docRoot,
    absolute: true,
  });

  return docPaths.map((docPath) => {
    const routePath = markdownPathToRoutePath(docPath, docRoot);
    const parts = routePath.split("/");
    const getContent = async () => {
      return await fsx.readFile(docPath, "utf-8");
    };
    return {
      name: parts[parts.length - 1] || "index",
      routePath: routePath,
      filePath: docPath,
      getContent,
      getFrontMatter: async () => {
        getFrontMatter(await getContent());
      },
    };
  });
};
