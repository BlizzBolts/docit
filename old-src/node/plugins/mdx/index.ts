import gfm from "remark-gfm";
import highlight from "rehype-highlight";
import type { Options } from "@mdx-js/rollup";
import mdxRollup from "@mdx-js/rollup";
import type { PluginOption } from "vite";
import frontMatter from "remark-frontmatter";
import emoji from "remark-emoji";
import slug from "rehype-slug";
import type {
  Options as RehypeAutoLinkHeadingsOptions,
} from "rehype-autolink-headings";
import rehypeAutolinkHeadings from "rehype-autolink-headings";
import type { ResolvedUserConfig } from "../../types.js";
import { live } from "./live.js";
import { api } from "./api.js";

export const getCompilerOptions = (config: ResolvedUserConfig): Options => {
  return {
    jsxRuntime: "classic",
    remarkPlugins: [gfm, frontMatter, emoji, [live, config], [api, config]],
    rehypePlugins: [
      highlight,
      slug,
      [
        rehypeAutolinkHeadings,
        {
          behavior: "wrap",
          properties: {
            className: "docit-heading-anchor",
          },
        } as RehypeAutoLinkHeadingsOptions,
      ],
    ],
    providerImportSource: "@mdx-js/react",
  };
};

export const mdx = async (
  config: ResolvedUserConfig
): Promise<PluginOption[]> => {
  return [mdxRollup(getCompilerOptions(config)) as PluginOption];
};
