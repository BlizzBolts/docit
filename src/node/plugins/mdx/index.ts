import gfm from "remark-gfm";
import highlight from "rehype-highlight";
import mdxRollup, { Options } from "@mdx-js/rollup";
import { ResolvedUserConfig } from "../../types.js";
import { PluginOption } from "vite";
import frontMatter from "remark-frontmatter";
import emoji from "remark-emoji";
import slug from "rehype-slug";
import { live } from "./live.js";
import { api } from "./api.js";
import rehypeAutolinkHeadings, {
  Options as RehypeAutoLinkHeadingsOptions,
} from "rehype-autolink-headings";

export const getCompilerOptions = (config: ResolvedUserConfig): Options => {
  return {
    jsxRuntime: "classic",
    remarkPlugins: [gfm, frontMatter, emoji, [live, config], [api, config]],
    rehypePlugins: [
      highlight,
      slug,
      [
        rehypeAutolinkHeadings,
        {
          behavior: "wrap",
          properties: {
            className: 'docit-heading-anchor'
          },
        } as RehypeAutoLinkHeadingsOptions,
      ],
    ],
    providerImportSource: "@mdx-js/react",
  };
};

export const mdx = async (
  config: ResolvedUserConfig
): Promise<PluginOption[]> => {
  return [mdxRollup(getCompilerOptions(config)) as PluginOption];
};
