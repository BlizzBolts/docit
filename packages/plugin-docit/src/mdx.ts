import type { PluginOption } from "vite";
import mdx from "@mdx-js/rollup";
import { type DocitConfig } from "@blizzbolts/docit-shared/node";
import gfm from "remark-gfm";
import emoji from "remark-emoji";
import frontmatter from "remark-frontmatter";
import mdxFrontmatter from "remark-mdx-frontmatter";

export const createMdxPlugin = async (config: DocitConfig): Promise<PluginOption[]> => {
  // const mdxPlugin: PluginOption = {
  //   name: "vite-plugin-docit-mdx",
  //   enforce: "pre",
  //   async transform(code, id, options) {
  //     return null;
  //   },
  // };

  return [
    // mdxPlugin,
    mdx({
      providerImportSource: "@mdx-js/react",
      remarkPlugins: [
        gfm,
        emoji,
        frontmatter,
        mdxFrontmatter,
      ],
    }),
  ];
};
