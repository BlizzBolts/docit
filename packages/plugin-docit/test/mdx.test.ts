import path from "node:path";
import fs from "fs-extra";
import { it } from "vitest";
import { describe } from "vitest";
import { compile } from "@mdx-js/mdx";
import remarkGfm from "remark-gfm";

describe("MDX", () => {
  it("should compile GFM", async ({ expect }) => {
    const fileContent = await fs.readFile(path.resolve(__dirname, "./fixtures/gfm.mdx"), "utf-8");
    const result = await compile(fileContent, {
      remarkPlugins: [remarkGfm],
    });

    expect(result).toMatchSnapshot();
  });

  it("should compile GFM full list", async ({ expect }) => {
    const fileContent = await fs.readFile(
      path.resolve(__dirname, "./fixtures/gfm-full.mdx"),
      "utf-8",
    );
    const result = await compile(fileContent, {
      remarkPlugins: [remarkGfm],
    });

    expect(result).toBeTruthy();
  });
});
