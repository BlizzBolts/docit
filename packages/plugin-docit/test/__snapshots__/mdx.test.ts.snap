// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MDX > should compile GFM 1`] = `
VFile {
  "cwd": "/Users/hao/spaces/projj/github.com/BlizzBolts/docit",
  "data": {},
  "history": [],
  "map": undefined,
  "messages": [],
  "value": "/*@jsxRuntime automatic @jsxImportSource react*/
import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from \\"react/jsx-runtime\\";
function _createMdxContent(props) {
  const _components = Object.assign({
    h1: \\"h1\\",
    h2: \\"h2\\",
    p: \\"p\\",
    a: \\"a\\",
    sup: \\"sup\\",
    del: \\"del\\",
    table: \\"table\\",
    thead: \\"thead\\",
    tr: \\"tr\\",
    th: \\"th\\",
    ul: \\"ul\\",
    li: \\"li\\",
    input: \\"input\\",
    section: \\"section\\",
    ol: \\"ol\\"
  }, props.components);
  return _jsxs(_Fragment, {
    children: [_jsx(_components.h1, {
      children: \\"GFM\\"
    }), \\"\\\\n\\", _jsx(_components.h2, {
      children: \\"Autolink literals\\"
    }), \\"\\\\n\\", _jsxs(_components.p, {
      children: [_jsx(_components.a, {
        href: \\"http://www.example.com\\",
        children: \\"www.example.com\\"
      }), \\", \\", _jsx(_components.a, {
        href: \\"https://example.com\\",
        children: \\"https://example.com\\"
      }), \\", and \\", _jsx(_components.a, {
        href: \\"mailto:contact@example.com\\",
        children: \\"contact@example.com\\"
      }), \\".\\"]
    }), \\"\\\\n\\", _jsx(_components.h2, {
      children: \\"Footnote\\"
    }), \\"\\\\n\\", _jsxs(_components.p, {
      children: [\\"A note\\", _jsx(_components.sup, {
        children: _jsx(_components.a, {
          href: \\"#user-content-fn-1\\",
          id: \\"user-content-fnref-1\\",
          \\"data-footnote-ref\\": true,
          \\"aria-describedby\\": \\"footnote-label\\",
          children: \\"1\\"
        })
      })]
    }), \\"\\\\n\\", _jsx(_components.h2, {
      children: \\"Strikethrough\\"
    }), \\"\\\\n\\", _jsxs(_components.p, {
      children: [_jsx(_components.del, {
        children: \\"one\\"
      }), \\" or \\", _jsx(_components.del, {
        children: \\"two\\"
      }), \\" tildes.\\"]
    }), \\"\\\\n\\", _jsx(_components.h2, {
      children: \\"Table\\"
    }), \\"\\\\n\\", _jsx(_components.table, {
      children: _jsx(_components.thead, {
        children: _jsxs(_components.tr, {
          children: [_jsx(_components.th, {
            children: \\"a\\"
          }), _jsx(_components.th, {
            align: \\"left\\",
            children: \\"b\\"
          }), _jsx(_components.th, {
            align: \\"right\\",
            children: \\"c\\"
          }), _jsx(_components.th, {
            align: \\"center\\",
            children: \\"d\\"
          })]
        })
      })
    }), \\"\\\\n\\", _jsx(_components.h2, {
      children: \\"Tasklist\\"
    }), \\"\\\\n\\", _jsxs(_components.ul, {
      className: \\"contains-task-list\\",
      children: [\\"\\\\n\\", _jsxs(_components.li, {
        className: \\"task-list-item\\",
        children: [_jsx(_components.input, {
          type: \\"checkbox\\",
          disabled: true
        }), \\" \\", \\"to do\\"]
      }), \\"\\\\n\\", _jsxs(_components.li, {
        className: \\"task-list-item\\",
        children: [_jsx(_components.input, {
          type: \\"checkbox\\",
          checked: true,
          disabled: true
        }), \\" \\", \\"done\\"]
      }), \\"\\\\n\\"]
    }), \\"\\\\n\\", _jsxs(_components.section, {
      \\"data-footnotes\\": true,
      className: \\"footnotes\\",
      children: [_jsx(_components.h2, {
        className: \\"sr-only\\",
        id: \\"footnote-label\\",
        children: \\"Footnotes\\"
      }), \\"\\\\n\\", _jsxs(_components.ol, {
        children: [\\"\\\\n\\", _jsxs(_components.li, {
          id: \\"user-content-fn-1\\",
          children: [\\"\\\\n\\", _jsxs(_components.p, {
            children: [\\"Big note. \\", _jsx(_components.a, {
              href: \\"#user-content-fnref-1\\",
              \\"data-footnote-backref\\": true,
              className: \\"data-footnote-backref\\",
              \\"aria-label\\": \\"Back to content\\",
              children: \\"â†©\\"
            })]
          }), \\"\\\\n\\"]
        }), \\"\\\\n\\"]
      }), \\"\\\\n\\"]
    })]
  });
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, Object.assign({}, props, {
    children: _jsx(_createMdxContent, props)
  })) : _createMdxContent(props);
}
export default MDXContent;
",
}
`;
